spock-specs/src/main/groovy/org/spockframework/specs/extension/SnapshotExtension.groovy

```diff
@@ -17,24 +21,63 @@ class SnapshotExtension implements IAnnotationDrivenExtension<Snapshot> {
   }
 
   @Override
-  void visitFieldAnnotation(Snapshot annotation, FieldInfo field) {
-    Assert.that(field.type.isAssignableFrom(Snapshotter), "Field must be of type Snapshotter")
-    field.parent.bottomSpec.addSetupInterceptor(new SnapshotInterceptor(annotation, field))
+  void visitSpecAnnotation(Snapshot annotation, SpecInfo specInfo) {
+    specInfo.features.each { visitFeatureAnnotation(annotation, it) }
+  }
+
+  @Override
+  void visitFeatureAnnotation(Snapshot annotation, FeatureInfo feature) {
+    def snapshotterFields = feature.spec.allFields.findAll {
+      ((it.type == Snapshotter) || (it.type.isAssignableFrom(Snapshotter) && (it.name == 'snapshotter'))) &&
+        !it.static && !it.shared
+    }
+    def snapshotterParameters = feature.featureMethod.parameters.findAll {
+      ((it.reflection.type == Snapshotter) || (it.reflection.type.isAssignableFrom(Snapshotter) && (it.name == 'snapshotter'))) &&
+        !(it.name in feature.dataVariables)
+    }
+    Assert.that(
+      (snapshotterFields.size() != 0) || (snapshotterParameters.size() != 0),
+      'There must be at least one Snapshotter parameter that is not a data variable or one Snapshotter field that is neither static nor @Shared'
+    )
+
+    feature.addIterationInterceptor(new SnapshotFieldInterceptor(annotation, snapshotterFields))
+    feature.featureMethod.addInterceptor(new SnapshotParameterInterceptor(annotation, snapshotterParameters))
```

Comment:
Why are the interceptors added even if there are now `fields` or `parameters` to inject?

Reply:
They are not.
If there is a `@Snapshot` annotation but neither a field nor parameter to inject the `Snapshotter`, execution fails due to the `Assert.that` before adding the interceptors, as it makes no sense to annotate with `@Snapshot` but then not having a `Snapshotter`.

Reply:
What I mean is that you are injecting both, if at least one of them is set.

Reply:
Oh, I see what you mean. I fixed it in the next-to-last commit before I anyway completely reverted everything. :-D